### DISCUSSION TITLE
strip_preamble with PDU

### DISCUSSION THREAD

**Discussion by richardetu** [duggabe/gr-control#11] [Q&A] [answered]
Hello, 

I watched your tutorial on packet transmission using BPSK, I documented myself at length with your YouTube video and what you posted on the gnuradio wiki.

I am experimenting with a USRP, and would like to transmit a png file for example, using your method. I'm stuck trying to decode the base64 file to the desired format. I can do this separately via a Python interpreter ("PyCharm").

In your video you use an Embeded Python Block which directly performs this function (PDU, EPB: Decode Packet). However, I was not able to find this block when I downloaded the reception GRC file, and I also did not find the python code on github to create the EPB myself.

Would you be able to provide on github the python code to put in an EPB or even better directly the block in a GRC file which would integrate the base64 decoding function?

I thank you in advance,
Regards,
![Capture d'écran 2024-02-11 002042](https://github.com/duggabe/gr-control/assets/159665877/100ada25-3933-48d7-83e8-327456725637)

**Comment by duggabe** [comment, marked as answer]
The current method of stripping the preamble, the postamble, and converting the Base64 back to the original file is to use the `strip_preamble.py` program as described in `https://wiki.gnuradio.org/index.php?title=File_transfer_using_Packet_and_BPSK#Stripping_the_Preamble_and_Filler`.

In my presentation I described the Embedded Python Block 'Decode Packet' as a "next step" to be done. I have not completed that project yet, but will resume working on it soon.

Using the current code, I have sent fairly large PNG and JPG images. Because of the Base64 conversions, there is no limit on the file content.

**Comment by duggabe** [comment]
Today I added `pkt_rcv_strip.grc` which integrates the `strip_preamble` function shown in the presentation. To use it, execute it in place of the `pkt_rcv.py` program.
Note however, files larger than 4kb tend to get data overruns and lost packets, as there is no flow control.

**Comment by richardetu** [comment]
I did some experiments this morning, with an SMA cable and antennas (on the same USRP, the antennas are not optimized). 

I transmitted a TXT file of 2.42 KB, a PNG file of 3.92 KB and an MP3 file of 140 KB. 

I also sent a 68KB image but there were very visible errors. While listening to the MP3 file I noticed some inconsistencies but nothing serious, very usable. The 3.92KB image was correct but the other had visible inconsistencies. 

The text file sometimes comes with character errors but nothing serious. I tried to send an mp4 file (120KB) but the errors make it unreadable. 

I think that depending on the type of file the errors are more or less critical because it did not pose a problem for the audio file but for other files of the same size it is unreadable.

**Comment by duggabe** [comment]
This implementation of stripping the preamble and converting the Base64 uses PDUs which are message-based. There is no guarantee that the messages will arrive in a timely manner or even in the original sequence.
A much more satisfactory solution would be to implement the function using streams. You are welcome to fork the repo and do that programming. If you get it working, please submit a Pull Request :)
